simSPI/linear_simulator/modules.py

```diff
@@ -0,0 +1,397 @@
+        If the path is not given or doesn't exist then the volume
+        is initialized with a cube.
+        """
+        if (
+            self.config.input_volume_path == ""
+            or os.path.exists(os.path.join(os.getcwd(), self.config.input_volume_path))
+            is False
+        ):
+
+            print(
+                "No input volume specified or the path doesn't exist. "
+                "Using cube as the default volume."
+            )
+            volume = init_cube(self.config.sidelen)
+        else:
+            with mrcfile.open(self.config.input_volume_path, "r") as m:
+                volume = torch.from_numpy(m.data.copy()).to(self.Projector.vol.device)
+
+        self.Projector.vol = volume
+
+
+"""Class to generate tomographic projection."""
+
+
+class Projector(torch.nn.Module):
+    """Class to generate tomographic projection.
+
+    Parameters
+    ----------
+    config: class
+        Class containing parameters of the Projector
+
+    """
+
+    def __init__(self, config):
+        super(Projector, self).__init__()
+
+        self.config = config
+        self.vol = torch.rand([self.config.sidelen] * 3, dtype=torch.float32)
+        lincoords = torch.linspace(
+            -1.0, 1.0, self.config.sidelen
+        )  # assume square volume
+        [X, Y, Z] = torch.meshgrid([lincoords, lincoords, lincoords])
+        coords = torch.stack([Y, X, Z], dim=-1)
+        self.register_buffer("vol_coords", coords.reshape(-1, 3))
+
+    def forward(self, rot_params, proj_axis=-1):
+        """Output the tomographic projection of the volume.
+
+        First rotate the volume and then sum it along an axis.
+
+        Parameters
+        ----------
+        rot_params: dict of type str to {tensor}
+            contains rotation matrix "rotmat" (chunks x 3 x 3) that is
+            used to rotate the volume
+        proj_axis: int
+            index along which summation is done of the rotated volume
+
+        Returns
+        -------
+        projection: tensor
+            Tensor containing tomographic projection
+            (chunks x 1 x sidelen x sidelen)
+        """
+        rotmat = rot_params["rotmat"]
+        batch_sz = rotmat.shape[0]
+        t = Rotate(rotmat, device=self.vol_coords.device)
+        rot_vol_coords = t.transform_points(self.vol_coords.repeat(batch_sz, 1, 1))
+
+        rot_vol = torch.nn.functional.grid_sample(
+            self.vol.repeat(batch_sz, 1, 1, 1, 1),  # = (Batch, C,D,H,W)
+            rot_vol_coords[:, None, None, :, :],
+            align_corners=True,
+        )
+        projection = torch.sum(
+            rot_vol.reshape(
+                batch_sz, self.config.sidelen, self.config.sidelen, self.config.sidelen
+            ),
+            dim=proj_axis,
+        )
+        projection = projection[
+            :, None, :, :
+        ]  # add a dummy channel (for consistency w/ img fmt)
+        # --> B,C,H,W
```

Comment:
This should be explained in the docstring instead (cf previous comment).

Reply:
Done. Changed in the new [PR](https://github.com/compSPI/simSPI/pull/23/).